test/integration/metrics/metrics_test.go

```diff
@@ -261,6 +263,90 @@ func TestAPIServerMetricsLabels(t *testing.T) {
 	}
 }
 
+func TestAPIServerMetricsLabelsWithAllowList(t *testing.T) {
+
+	allowMetricLabels := []struct {
+		metricName  string
+		labelName   model.LabelName
+		allowValues model.LabelValues
+		isHistogram bool
+	}{
+		{
+			// CounterVec metric
+			metricName:  "apiserver_request_total",
+			labelName:   "code",
+			allowValues: model.LabelValues{"201", "500"},
+		},
+		{
+			// GaugeVec metric
+			metricName:  "apiserver_current_inflight_requests",
+			labelName:   "request_kind",
+			allowValues: model.LabelValues{"mutating"},
+		},
+		{
+			// Histogram metric
+			metricName:  "apiserver_request_duration_seconds",
+			labelName:   "verb",
+			allowValues: model.LabelValues{"POST", "LIST"},
+			isHistogram: true,
+		},
+	}
+
+	// Assemble the allow-metric-labels flag.
+	var allowMetricLabelFlagStrs []string
+	for _, allowMetricLabel := range allowMetricLabels {
+		var allowValuesStr []string
+		for _, allowValue := range allowMetricLabel.allowValues {
+			allowValuesStr = append(allowValuesStr, string(allowValue))
+		}
+		allowMetricLabelFlagStrs = append(allowMetricLabelFlagStrs, fmt.Sprintf("\"%s,%s=%s\"", allowMetricLabel.metricName, allowMetricLabel.labelName, strings.Join(allowValuesStr, ",")))
+	}
+	allowMetricLabelsFlag := "--allow-metric-labels=" + strings.Join(allowMetricLabelFlagStrs, ",")
+
+	testServerFlags := framework.DefaultTestServerFlags()
+	testServerFlags = append(testServerFlags, allowMetricLabelsFlag)
+
+	// KUBE_APISERVER_SERVE_REMOVED_APIS_FOR_ONE_RELEASE allows for APIs pending removal to not block tests
+	t.Setenv("KUBE_APISERVER_SERVE_REMOVED_APIS_FOR_ONE_RELEASE", "true")
+	s := kubeapiservertesting.StartTestServerOrDie(t, nil, testServerFlags, framework.SharedEtcd())
+	defer s.TearDownFn()
+
+	metrics, err := scrapeMetrics(s)
+	if err != nil {
+		t.Fatal(err)
+	}
+	for _, allowMetric := range allowMetricLabels {
+		metricName := allowMetric.metricName
+		if allowMetric.isHistogram {
+			metricName = metricName + "_sum"
+		}
+		if samples, found := metrics[metricName]; !found {
+			t.Errorf("metric %q not found", metricName)
+		} else {
+			for _, sample := range samples {
+				if value, ok := sample.Metric[allowMetric.labelName]; ok {
+					if !slices.Contains(allowMetric.allowValues, value) && value != "unexpected" {
+						t.Errorf("value %q is not allowed for label %q", value, allowMetric.labelName)
```

Comment:
should this be t.Fatalf() ? Same on L#324, L#338 ? That will also help with removing the subsequent `else` blocks

Reply:
Updated. Thanks!